src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -1552,391 +1933,247 @@ fn slice_pat_covered_by_const<'tcx>(
+            let included = (order_from != Ordering::Less)
+                && ((order_to == Ordering::Less)
+                    || (pat_end == ctor_end && order_to == Ordering::Equal));
+            if included { Some(PatStack::default()) } else { None }
         }
-        Single => Ok(true),
-        _ => bug!(),
+        _ => bug!("`constructor_intersects_pattern` called with {:?}", ctor),
     }
 }
 
 fn patterns_for_variant<'p, 'tcx>(
     subpatterns: &'p [FieldPat<'tcx>],
-    wild_patterns: &[&'p Pat<'tcx>])
-    -> SmallVec<[&'p Pat<'tcx>; 2]>
-{
-    let mut result = SmallVec::from_slice(wild_patterns);
+    ctor_wild_subpatterns: &[&'p Pat<'tcx>],
+) -> PatStack<'p, 'tcx> {
+    let mut result = SmallVec::from_slice(ctor_wild_subpatterns);
 
     for subpat in subpatterns {
         result[subpat.field.index()] = &subpat.pattern;
     }
 
-    debug!("patterns_for_variant({:#?}, {:#?}) = {:#?}", subpatterns, wild_patterns, result);
-    result
+    debug!(
+        "patterns_for_variant({:#?}, {:#?}) = {:#?}",
+        subpatterns, ctor_wild_subpatterns, result
+    );
+    PatStack::from_vec(result)
 }
 
-/// This is the main specialization step. It expands the first pattern in the given row
-/// into `arity` patterns based on the constructor. For most patterns, the step is trivial,
-/// for instance tuple patterns are flattened and box patterns expand into their inner pattern.
+/// This is the main specialization step. It expands the pattern into `arity` patterns based on the
+/// constructor. For most patterns, the step is trivial, for instance tuple patterns are flattened
+/// and box patterns expand into their inner pattern. Returns vec![] if the pattern does not have
+/// the given constructor. See the top of the file for details.
 ///
-/// OTOH, slice patterns with a subslice pattern (..tail) can be expanded into multiple
-/// different patterns.
 /// Structure patterns with a partial wild pattern (Foo { a: 42, .. }) have their missing
 /// fields filled with wild patterns.
-fn specialize<'p, 'a: 'p, 'tcx>(
-    cx: &mut MatchCheckCtxt<'a, 'tcx>,
-    r: &[&'p Pat<'tcx>],
+fn specialize_one_pattern<'p, 'a: 'p, 'p2: 'p, 'tcx>(
+    cx: &MatchCheckCtxt<'a, 'tcx>,
+    mut pat: &'p2 Pat<'tcx>,
     constructor: &Constructor<'tcx>,
-    wild_patterns: &[&'p Pat<'tcx>],
-) -> Option<SmallVec<[&'p Pat<'tcx>; 2]>> {
-    let pat = &r[0];
+    ctor_wild_subpatterns: &[&'p Pat<'tcx>],
+) -> SmallVec<[PatStack<'p, 'tcx>; 1]> {
+    while let PatKind::AscribeUserType { ref subpattern, .. } = *pat.kind {
+        pat = subpattern;
+    }
 
-    let head = match *pat.kind {
-        PatKind::AscribeUserType { ref subpattern, .. } => {
-            specialize(cx, ::std::slice::from_ref(&subpattern), constructor, wild_patterns)
-        }
+    if let Wildcard | MissingConstructors(_) = constructor {
+        // If `constructor` is `Wildcard`: either there were only wildcards in the first component
```

Comment:
I think this should be a note on the top-level comment with a reference to it. Say, replace this comment with `see (MISSING-CTOR)`, and add the following to the top-level comment:

```rust
/// # Handling of missing constructors (MISSING-CTOR)
/// 
/// This algorithm can be seen to essentially explore the full decision
/// tree of patterns, except in the case where it "takes a shortcut"
/// find a "missing" constructor - that's it, a constructor that is not
/// covered by any of the non-wildcard patterns
/// 
/// For example, in the following case:
/// ```rust
/// match x {
///     (_, true) => {}
///     (None, false) => {}
/// }
/// ```
///
/// The algorithm proceeds as follows:
/// ```
/// M = [ [(_, true)],
///       [(None, false)]]
/// p = [_]
/// -- expand tuple (ctor = Variant)
///   M = [ [_, true],
///         [None, false]]
///   p = [_, _]
///   -- `Some(_)` is a missing ctor, dropping all the non-wildcard arms
///     M = [ [true]]
///     p = [_]
///     -- `false` is a possible match
///       M = []
///       p = []
///     return "[false]"
///   return "[Some(_), false]"
/// return "[(Some(_), false)]"
/// ```
/// 
/// Once it finds that `Some(_)` is a missing constructor, it does not
/// need to look any further - any witness using a non-missing constructor
/// can be transformed to a witness using the missing constructor - and
/// therefore it does not try to look for witnesses involving the
/// other constructors - in this case, the `(None, true)` witness
/// (which can be "transformed" to `(Some(_), true)`).
/// 
/// In the code, missing constructors are represented by the `Wildcard` and
/// `MissingConstructors` variants of `Constructor`, with the difference
/// between them being down to error reporting: `MissingConstructors`
/// "remembers" the set of constructors it contains for error reporting
/// (so we can show the `... not covered` error message), while `Wildcard`
/// doesn't.
/// 
/// Therefore, `Wildcard` is used in cases where the exact constructor
/// doesn't matter - either where the head column of the matrix contains
/// only wildcards (and therefore, *every* constructor will work) or
/// when the enum is `#[non_exhaustive]`, and therefore from a user POV
/// there can always assumed to be an "fresh" constructor that will be
/// useful for the witness, while `MissingConstructors` is used in
/// the other cases.
```

Reply:
Thanks for the very detailed comment !